daffodil-lib/src/main/scala/org/apache/daffodil/util/Delay.scala

```diff
@@ -28,17 +28,45 @@ package org.apache.daffodil.util
  * as Delay objects.
  *
  * For more info, this is used in the IncludeImport stuff in DSOM.
+ *
+ * This is NOT serializable, and shouldn't be made so.
  */
 
-class Delay[T] private (v: => T) {
-  lazy val value = v
-  private val hasValue: Boolean = false
+final class Delay[T] private (private var box: Delay.Box[T]) {
+  //
+  // This trick of taking another object on a var, and
+  // then clobbering the var after we demand the value
+  // eliminates holding onto a bunch of objects due
+  // to closures for the pass-by-name arguments.
+  //
+  lazy val value: T = {
+    val v = box.value
+    box = null // throws away box, which allows GC of all closures, etc.
+    v
+  }
+
+  def hasValue = box ne null
```

Comment:
Surprised nothing actually calls hasValue? It's pretty clearly correct though. Is it worth adding a unit test to ensure this works as expected and doesn't ever have a regression?

Reply:
Really used only by the toString method, to tell it whether to access the value or not.